cmd/pkg/baseline/generator_oscal.go

```diff
@@ -0,0 +1,109 @@
+// SPDX-FileCopyrightText: Copyright 2025 The OSPS Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package baseline
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	oscal "github.com/defenseunicorns/go-oscal/src/types/oscal-1-1-3"
+	"github.com/ossf/security-baseline/pkg/types"
+)
+
+const (
+	VersionOSPS = "devel"
+	controlHREF = "https://baseline.openssf.org/versions/%s#%s"
+	catalogUUID = "8c222a23-fc7e-4ad8-b6dd-289014f07a9f"
+)
+
+func (g *Generator) ExportOSCAL(b *types.Baseline, w io.Writer) error {
+	n := time.Now()
+	catalog := oscal.Catalog{
+		UUID:   catalogUUID,
+		Groups: nil,
+		Metadata: oscal.Metadata{
+			LastModified: n,
+			Links: &[]oscal.Link{
+				{
+					Href: fmt.Sprintf(controlHREF, VersionOSPS, ""),
+				},
+			},
+			OscalVersion: "1.1.3",
+			Published:    &n,
+			Title:        "Open Source Project Security Baseline",
+			Version:      VersionOSPS,
+		},
+	}
+
+	catalogGroups := []oscal.Group{}
+
+	for code, cat := range b.Categories {
+		group := oscal.Group{
+			Class:    "OSPS",
+			Controls: nil,
+			ID:       code,
+			Title:    cat.Description,
+		}
+
+		controls := []oscal.Control{}
+
+		for _, criteria := range cat.Criteria {
+			newCtl := oscal.Control{
+				Class: criteria.Category,
+				ID:    strings.TrimPrefix(criteria.ID, "OSPS-"),
+				Links: &[]oscal.Link{
+					{
+						Href: fmt.Sprintf(controlHREF, VersionOSPS, strings.ToLower(criteria.ID)),
+						Rel:  "reference",
+					},
+				},
+				Props: &[]oscal.Property{
+					{
+						Name:  "maturity-level",
+						UUID:  catalogUUID,
+						Value: fmt.Sprintf("%d", criteria.MaturityLevel),
+					},
+				},
+				Parts: &[]oscal.Part{
+					{
+						ID:    strings.TrimPrefix(criteria.ID, "OSPS-") + "_details",
+						Name:  "details",
+						Prose: strings.TrimSpace(criteria.Details),
+					},
+					{
+						ID:    strings.TrimPrefix(criteria.ID, "OSPS-") + "_rationale",
+						Name:  "rationale",
+						Prose: strings.TrimSpace(criteria.Rationale),
+					},
+				},
+				Title: strings.TrimSpace(criteria.CriterionText),
+			}
+
+			if criteria.Implementation != "" {
+				pts := append(*newCtl.Parts, oscal.Part{
+					ID:    strings.TrimPrefix(criteria.ID, "OSPS-") + "_implementation",
+					Name:  "implementation",
+					Prose: strings.TrimSpace(criteria.Implementation),
+				})
+				newCtl.Parts = &pts
+			}
+
+			controls = append(controls, newCtl)
+		}
+
+		group.Controls = &controls
+		catalogGroups = append(catalogGroups, group)
+	}
+	catalog.Groups = &catalogGroups
+
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	if err := enc.Encode(catalog); err != nil {
```

Comment:
This will produce the file for all of the content within the catalog model - but if we look at a [catalog example](https://github.com/usnistgov/oscal-content/blob/main/examples/catalog/json/basic-catalog.json#L2) we need the `catalog` object key included for validity of the model. 

This can be done a number of ways. Typically we use the [complete schema](https://github.com/defenseunicorns/go-oscal/blob/v0.6.2/src/types/oscal-1-1-3/types.go#L23) to reference the catalog and encode the complete schema. 

go-oscal has a CLI with a `validate` command that can perform schema validation of models if that is helpful.

Reply:
Ah thanks for the tip! I've wrapped the doc into a "catalog" label and fixed the validation:

```
go run . validate --input-file ../ossf/security-baseline/cmd/oscal.json  
go-oscal: 2025/02/20 19:05:13 Successfully validated ../ossf/security-baseline/cmd/oscal.json is valid OSCAL version 1.1.3 catalog
```